modules/backend/src/main/scala/database.operations.scala

```diff
@@ -5,73 +5,72 @@ import cats.effect.IO
 import spec.*
 
 import cats.*, cats.data.*, cats.implicits.*
-import doobie.*, doobie.implicits.*
+import skunk.{Codec, Query}
+import skunk.implicits.*
+import skunk.codec.all.*
+
 import smithy4s.Newtype
+import skunk.data.Completion
+import skunk.Command
 
 sealed trait SqlOp[I, O]
 
-sealed abstract class SqlQuery[I, O](val input: I, val out: Query0[O])
+sealed abstract class SqlQuery[I, O](val input: I, val out: Query[I, O])
     extends SqlOp[I, O]
-sealed abstract class SqlUpdate[I](val input: I, val out: Update0)
-    extends SqlOp[I, Int]
-
-private def ntGet[T: Get](nt: Newtype[T]): Get[nt.Type] =
-  Get[T].map(nt.apply)
-
-private def ntPut[T: Put](nt: Newtype[T]): Put[nt.Type] =
-  Put[T].contramap[nt.Type](_.value)
 
-given Get[Key]   = ntGet(Key)
-given Get[Value] = ntGet(Value)
+sealed abstract class SqlUpdate[I](val input: I, val out: Command[I])
+    extends SqlOp[I, Int]
 
-given Put[Key]   = ntPut(Key)
-given Put[Value] = ntPut(Value)
+def newtypeCodec[A](nt: Newtype[A], underlying: Codec[A]): Codec[nt.Type] =
+  underlying.imap(nt.apply)(_.value)
+val keyCodec   = newtypeCodec(Key, varchar(50))
+val valueCodec = newtypeCodec(Value, int4)
+val pairCodec  = (keyCodec *: valueCodec).to[Pair]
 
 object operations:
 
   case class Get(key: Key)
       extends SqlQuery(
         key,
-        sql"select value from examples where key = $key"
-          .query[Int]
-          .map(Value.apply)
+        sql"select value from examples where key = ${newtypeCodec(Key, varchar(50))}"
+          .query(valueCodec)
       )
 
   case object GetAll
       extends SqlQuery(
-        (),
+        skunk.Void,
         sql"select key, value from examples order by key"
-          .query[(Key, Value)]
-          .map(Pair.apply)
+          .query(pairCodec)
       )
 
-  case class Create(key: Key, value: Option[Value])
+  case class Create(key: Key, value: Value)
       extends SqlUpdate(
-        key,
-        sql"insert into examples (key, value) values ($key, ${value.map(_.value).getOrElse(0)})".update
```

Comment:
Was unsure the database operation is the best place for `getOrElse`. Alternatively you can `contramap`.